File: clients/ruby/model/mine.rb
require_relative 'vec2_double'
require_relative 'vec2_double'
require_relative 'mine_state'
require_relative 'explosion_params'
class Mine
    attr_accessor :player_id
    attr_accessor :position
    attr_accessor :size
    attr_accessor :state
    attr_accessor :timer
    attr_accessor :trigger_radius
    attr_accessor :explosion_params
    def initialize(player_id, position, size, state, timer, trigger_radius, explosion_params)
        @player_id = player_id
        @position = position
        @size = size
        @state = state
        @timer = timer
        @trigger_radius = trigger_radius
        @explosion_params = explosion_params
    end
    def self.read_from(stream)
        player_id = stream.read_int()
        position = Vec2Double.read_from(stream)
        size = Vec2Double.read_from(stream)
        state = stream.read_int()
        if state < 0 || state > 4
            raise "Unexpected discriminant value"
        end
        if stream.read_bool()
            timer = stream.read_double()
        else
            timer = nil
        end
        trigger_radius = stream.read_double()
        explosion_params = ExplosionParams.read_from(stream)
        Mine.new(player_id, position, size, state, timer, trigger_radius, explosion_params)
    end
    def write_to(stream)
        stream.write_int(@player_id)
        @position.write_to(stream)
        @size.write_to(stream)
        stream.write_int(@state)
        if @timer.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            stream.write_double(@timer)
        end
        stream.write_double(@trigger_radius)
        @explosion_params.write_to(stream)
    end
end

File: clients/python/model/mine.py
from .vec2_double import Vec2Double
from .vec2_double import Vec2Double
from .mine_state import MineState
from .explosion_params import ExplosionParams
class Mine:
    def __init__(self, player_id, position, size, state, timer, trigger_radius, explosion_params):
        self.player_id = player_id
        self.position = position
        self.size = size
        self.state = state
        self.timer = timer
        self.trigger_radius = trigger_radius
        self.explosion_params = explosion_params
    @staticmethod
    def read_from(stream):
        player_id = stream.read_int()
        position = Vec2Double.read_from(stream)
        size = Vec2Double.read_from(stream)
        state = MineState(stream.read_int())
        if stream.read_bool():
            timer = stream.read_double()
        else:
            timer = None
        trigger_radius = stream.read_double()
        explosion_params = ExplosionParams.read_from(stream)
        return Mine(player_id, position, size, state, timer, trigger_radius, explosion_params)
    def write_to(self, stream):
        stream.write_int(self.player_id)
        self.position.write_to(stream)
        self.size.write_to(stream)
        stream.write_int(self.state)
        if self.timer is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            stream.write_double(self.timer)
        stream.write_double(self.trigger_radius)
        self.explosion_params.write_to(stream)
    def __repr__(self):
        return "Mine(" + \
            repr(self.player_id) + "," + \
            repr(self.position) + "," + \
            repr(self.size) + "," + \
            repr(self.state) + "," + \
            repr(self.timer) + "," + \
            repr(self.trigger_radius) + "," + \
            repr(self.explosion_params) + \
            ")"
